File: packages/go/nurseandrea/tracing/exporter.go
package tracing

import (
	"bytes"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"sync"
	"time"

	"github.com/narteyb/nurse-andrea-sdks/packages/go/nurseandrea"
)

const (
	batchSize     = 100
	flushInterval = 5 * time.Second
)

type OTLPSpan struct {
	TraceID            string          `json:"traceId"`
	SpanID             string          `json:"spanId"`
	ParentSpanID       string          `json:"parentSpanId"`
	Name               string          `json:"name"`
	Kind               int             `json:"kind"`
	StartTimeUnixNano  string          `json:"startTimeUnixNano"`
	EndTimeUnixNano    string          `json:"endTimeUnixNano"`
	Status             OTLPStatus      `json:"status"`
	Attributes         []OTLPAttribute `json:"attributes"`
	Events             []interface{}   `json:"events"`
}

type OTLPStatus struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

type OTLPAttribute struct {
	Key   string                 `json:"key"`
	Value map[string]interface{} `json:"value"`
}

var (
	mu      sync.Mutex
	queue   []OTLPSpan
	started bool
	stopCh  chan struct{}
)

func GenerateTraceID() string {
	b := make([]byte, 16)
	rand.Read(b)
	return hex.EncodeToString(b)
}

func GenerateSpanID() string {
	b := make([]byte, 8)
	rand.Read(b)
	return hex.EncodeToString(b)
}

func EnqueueSpan(span OTLPSpan) {
	if !nurseandrea.IsEnabled() {
		return
	}
	mu.Lock()
	queue = append(queue, span)
	shouldFlush := len(queue) >= batchSize
	mu.Unlock()
	if shouldFlush {
		go flush()
	}
}

func init() {
	nurseandrea.SetTracingStartFunc(Start)
}

func Start() {
	if started {
		return
	}
	started = true
	stopCh = make(chan struct{})
	go flushLoop()
}

func Stop() {
	if !started {
		return
	}
	close(stopCh)
	flush()
}

func flushLoop() {
	ticker := time.NewTicker(flushInterval)
	defer ticker.Stop()
	for {
		select {
		case <-ticker.C:
			flush()
		case <-stopCh:
			return
		}
	}
}

func flush() {
	mu.Lock()
	if len(queue) == 0 {
		mu.Unlock()
		return
	}
	spans := make([]OTLPSpan, len(queue))
	copy(spans, queue)
	queue = queue[:0]
	mu.Unlock()

	cfg := nurseandrea.GetConfig()
	payload := map[string]interface{}{
		"resourceSpans": []map[string]interface{}{{
			"resource": map[string]interface{}{
				"attributes": []map[string]interface{}{
					{"key": "service.name", "value": map[string]string{"stringValue": cfg.ServiceName}},
				},
			},
			"scopeSpans": []map[string]interface{}{
				{"spans": spans},
			},
		}},
	}

	body, err := json.Marshal(payload)
	if err != nil {
		fmt.Fprintf(os.Stderr, "[NurseAndrea] Trace marshal error: %v\n", err)
		return
	}

	url := cfg.Host + "/api/v1/traces"
	req, err := http.NewRequest("POST", url, bytes.NewReader(body))
	if err != nil {
		fmt.Fprintf(os.Stderr, "[NurseAndrea] Trace request error: %v\n", err)
		return
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+cfg.Token)

	resp, err := (&http.Client{Timeout: 5 * time.Second}).Do(req)
	if err != nil {
		fmt.Fprintf(os.Stderr, "[NurseAndrea] Trace export failed: %v\n", err)
		return
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		fmt.Fprintf(os.Stderr, "[NurseAndrea] Trace export → %d\n", resp.StatusCode)
	}
}

// MakeServerSpan builds an OTLP span for an HTTP server request.
func MakeServerSpan(method, path string, statusCode int, startNs, endNs int64, serviceName string) OTLPSpan {
	code := 1
	msg := ""
	if statusCode >= 500 {
		code = 2
		msg = fmt.Sprintf("HTTP %d", statusCode)
	}
	return OTLPSpan{
		TraceID:           GenerateTraceID(),
		SpanID:            GenerateSpanID(),
		ParentSpanID:      "",
		Name:              fmt.Sprintf("%s %s", method, path),
		Kind:              2,
		StartTimeUnixNano: fmt.Sprintf("%d", startNs),
		EndTimeUnixNano:   fmt.Sprintf("%d", endNs),
		Status:            OTLPStatus{Code: code, Message: msg},
		Attributes: []OTLPAttribute{
			{Key: "http.method", Value: map[string]interface{}{"stringValue": method}},
			{Key: "http.url", Value: map[string]interface{}{"stringValue": path}},
			{Key: "http.status_code", Value: map[string]interface{}{"intValue": statusCode}},
		},
		Events: []interface{}{},
	}
}

File: packages/node/src/tracing/exporter.ts
import { getConfig, isEnabled } from "../configuration"
import crypto from "crypto"

interface SpanData {
  traceId: string
  spanId: string
  parentSpanId: string
  name: string
  kind: number
  startTimeUnixNano: string
  endTimeUnixNano: string
  status: { code: number; message: string }
  attributes: Array<{ key: string; value: Record<string, unknown> }>
  events: unknown[]
}

const queue: SpanData[] = []
let timer: ReturnType<typeof setInterval> | null = null

export function generateTraceId(): string {
  return crypto.randomBytes(16).toString("hex")
}

export function generateSpanId(): string {
  return crypto.randomBytes(8).toString("hex")
}

export function enqueueSpan(span: SpanData): void {
  if (!isEnabled()) return
  queue.push(span)
  if (queue.length >= 100) flush()
}

export function startTraceExporter(): void {
  if (timer) return
  timer = setInterval(flush, 5000)
  if (timer.unref) timer.unref()
}

export function stopTraceExporter(): void {
  if (timer) {
    clearInterval(timer)
    timer = null
  }
  flush()
}

async function flush(): Promise<void> {
  if (!isEnabled() || queue.length === 0) return
  const spans = queue.splice(0)
  const config = getConfig()

  const payload = {
    resourceSpans: [{
      resource: {
        attributes: [
          { key: "service.name", value: { stringValue: config.serviceName } }
        ]
      },
      scopeSpans: [{ spans }]
    }]
  }

  const url = `${config.host.replace(/\/$/, "")}/api/v1/traces`
  try {
    const res = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type":              "application/json",
        "Authorization":             `Bearer ${config.orgToken}`,
        "X-NurseAndrea-Workspace":   config.workspaceSlug,
        "X-NurseAndrea-Environment": config.environment,
      },
      body: JSON.stringify(payload),
    })
    if (!res.ok) {
      process.stderr.write(`[NurseAndrea] Trace export → ${res.status}\n`)
    }
  } catch (err) {
    process.stderr.write(`[NurseAndrea] Trace export failed: ${(err as Error).message}\n`)
  }
}
